cmd/rpc/sock.go

```diff
@@ -0,0 +1,455 @@
+	afterRCUpdate func(info lib.RootChainInfo) // callback after the root chain info update
+	upgrader      websocket.Upgrader           // upgrade http connection to ws
+	log           lib.LoggerI                  // stout log
+}
+
+// NewRCManager() constructs a new instance of a RCManager
+func NewRCManager(controller *controller.Controller, c lib.Config, l lib.LoggerI) (man *RCManager) {
+	// create the manager
+	man = &RCManager{
+		c:             c,
+		subscriptions: make(map[uint64]*RCSubscription),
+		subscribers:   make(map[uint64]*RCSubscriber),
+		l:             controller.Mutex,
+		afterRCUpdate: controller.UpdateRootChainInfo,
+		upgrader:      websocket.Upgrader{CheckOrigin: func(r *http.Request) bool { return true }},
+		log:           l,
+	}
+	// set the manager in the controller
+	controller.RCManager = man
+	// exit
+	return
+}
+
+// Start() attempts to establish a websocket connection with each root chain
+func (r *RCManager) Start() {
+	// for each rc in the config
+	for _, rc := range r.c.RootChain {
+		// dial each root chain
+		r.NewSubscription(rc)
+	}
+}
+
+// Publish() writes the root-chain info to each client
+func (r *RCManager) Publish(chainId uint64, info *lib.RootChainInfo) {
+	// convert the root-chain info to bytes
+	protoBytes, err := lib.Marshal(info)
+	if err != nil {
+		return
+	}
+	// for each ws client
+	for _, subscriber := range r.subscribers {
+		// skip if not this chain id
+		if subscriber.chainId != chainId {
+			continue
+		}
+		// publish to each client
+		if e := subscriber.conn.WriteMessage(websocket.BinaryMessage, protoBytes); e != nil {
+			subscriber.Stop(e)
+		}
+	}
+}
+
+// ChainIds() returns a list of chainIds for subscribers
+func (r *RCManager) ChainIds() (list []uint64) {
+	// de-duplicate the results
+	deDupe := lib.NewDeDuplicator[uint64]()
+	// for each client
+	for _, client := range r.subscribers {
+		// if the client chain id isn't empty and not duplicate
+		if client.chainId != 0 && !deDupe.Found(client.chainId) {
+			list = append(list, client.chainId)
+		}
+	}
+	return
+}
+
+// GetHeight() returns the height from the root-chain
+func (r *RCManager) GetHeight(rootChainId uint64) uint64 {
+	// check the map to see if the info exists
+	if sub, found := r.subscriptions[rootChainId]; found {
+		// exit with the height of the root-chain-info
+		return sub.Info.Height
+	}
+	return 0
+}
+
+// GetRootChainInfo() retrieves the root chain info from the root chain 'on-demain'
+func (r *RCManager) GetRootChainInfo(rootChainId, chainId uint64) (rootChainInfo *lib.RootChainInfo, err lib.ErrorI) {
+	// lock for thread safety
+	r.l.Lock()
+	defer r.l.Unlock()
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return nil, lib.ErrNotSubscribed()
+	}
+	// get the info
+	info, err := sub.RootChainInfo(0, chainId)
+	if err != nil {
+		return nil, err
+	}
+	// update the info
+	sub.Info = *info
```

Comment:
generated rpc types contain internally a mutex too and this is also being copied by value which is not recommended `RCSubscription.Info` should be a pointer reference to avoid unexpected behavior with the lock

Reply:
I don't think it has a mutex